### PR TITLE
xrange() was removed in Python 3 in favor of range()

### PR DESCRIPTION
**Pull request by cclauss**
#### What does this implement/fix?

Remove all instances of __from scipy.\_lib.six import xrange__ and
convert all calls to __xrange()__ into calls to __range()__.

### CODE REVIEW

File: scipy/stats/_binned_statistic.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 
 def binned_statistic(x, values, statistic='mean',
-                     bins=10, range=None):
+                     bins=10, _range=None):
```
Comment (ev-br) [inline scipy/stats/_binned_statistic.py]: This breaks backwards compatibility since parameter names are effectively public API. Can you please undo this change, here and elsewhere.

Comment (cclauss) [reply scipy/stats/_binned_statistic.py]: Unfortunately __range__ is a Python builtin in both Python 2 and Python 3.  That makes it a poor choice for a variable name.

Comment (ev-br) [reply scipy/stats/_binned_statistic.py]: Indeed. But that ship has likely sailed. 

This is not to say we never rename poorly named arguments. We can, if the benefit outweights a possible disruption downstream. Which I don't think it does, in this particular case.

**Comment by ev-br** [comment]
Looks reasonable modulo a possible backcompat break. I'd suggest to consider it separately and remove from this PR. 

Also, I've labelled it Py3K (after numpy), so that we can group these rip-out-py2 PRs.

File: scipy/stats/_binned_statistic.py (hunk 1 above)
Comment (cclauss) [reply scipy/stats/_binned_statistic.py]: What do you want to call the builtin __range()__??
1. xrange()?
2. builtin_range()?
3. range_builtin()?
4. range_function()?
5. other?

Comment (perimosocordiae) [reply scipy/stats/_binned_statistic.py]: In this function we don't call the builtin `range`, so there's no conflict and nothing needs to be changed.

Comment (cclauss) [reply scipy/stats/_binned_statistic.py]: The problem comes in __binned_statistic_dd()__ and ___bin_edges()__ that use both the builtin __range()__ and the function parameter __range__.

Comment (ev-br) [reply scipy/stats/_binned_statistic.py]: How about an explicit `import builins; builtins.range(...)`?

Comment (cclauss) [reply scipy/stats/_binned_statistic.py]: Nice!  (For a separate PR.)

File: scipy/interpolate/polyint.py [hunk 2]
```diff
@@ -571,7 +570,7 @@ def add_xi(self, xi, yi=None):
         old_wi = self.wi
         self.wi = np.zeros(self.n)
         self.wi[:old_n] = old_wi
-        for j in xrange(old_n,self.n):
+        for j in range(old_n,self.n):
```
Comment (andyfaff) [inline scipy/interpolate/polyint.py]: nit, space after the comma

File: scipy/interpolate/tests/test_polyint.py [hunk 3]
```diff
@@ -175,27 +173,27 @@ def test_scalar(self):
...
         m = 10
         r = P.derivatives(self.test_xs,m)
-        for i in xrange(m):
+        for i in range(m):
             assert_almost_equal(P.derivative(self.test_xs,i),r[i])
 
     def test_high_derivative(self):
         P = KroghInterpolator(self.xs,self.ys)
-        for i in xrange(len(self.xs),2*len(self.xs)):
+        for i in range(len(self.xs),2*len(self.xs)):
```
Comment (andyfaff) [inline scipy/interpolate/tests/test_polyint.py]: nit, space after the comma

File: scipy/interpolate/polyint.py (hunk 2 above)
Comment (cclauss) [reply scipy/interpolate/polyint.py]: In a separate pull request, just pass the whole codebase thru psf/black.

File: scipy/sparse/tests/test_base.py [hunk 4]
```diff
@@ -2570,8 +2570,8 @@ def test_set_slice(self):
                     B[a,b] = 10*i + 1000*(j+1)
                     assert_array_equal(A.todense(), B, repr((a, b)))
 
-            A[0, 1:10:2] = xrange(1,10,2)
-            B[0, 1:10:2] = xrange(1,10,2)
+            A[0, 1:10:2] = range(1,10,2)
```
Comment (andyfaff) [inline scipy/sparse/tests/test_base.py]: nit, spaces after commas

**Review by andyfaff — APPROVED** [review]
A few nits where a space after a comma could be added. @cclauss, could you put those spaces in, and squash down to a single commit. Then the PR is mergeable. Thank you for the PR.

File: scipy/interpolate/polyint.py (hunk 2 above)
Comment (andyfaff) [reply scipy/interpolate/polyint.py]: In an ideal world it would be nice to have that. However, this would cause havoc to the codebase as there would be so many files touched, creating issues for in progress PR's, churn, etc.  Making all those changes to the entire codebase would cause more harm than good. It's better to make such changes as coincidental alterations (such as this) are made.
